homeassistant/components/energy/sensor.py

```diff
@@ -327,6 +341,9 @@ async def async_added_to_hass(self) -> None:
 
         self._attr_name = f"{name} {self._adapter.name_suffix}"
 
+        if metadata := statistics.get_metadata(self.hass, self.entity_id):
+            self._currency = metadata["unit_of_measurement"]
```

Comment:
There's a race here, there may be old states which are not yet in statistics meta data, for example:
1. Configure the energy cost sensor with a manually configured price entity
2. There is at least one state change of the energy sensor
3. Update the configuration of the price entity to a different currency and restart Home Assistant
4. There is no metadata, so self._currency will be `None`
5. When statistics runs, there will be conflicting UoM

Reply:
Fixed by getting past unit from stored state instead